chain/pruned_block_dispatcher.go

```diff
@@ -550,6 +563,10 @@ func (d *PrunedBlockDispatcher) newRequest(blocks []*chainhash.Hash) (
 		} else {
 			log.Debugf("Received new request for pending query of "+
 				"block %v", *block)
+
+			d.blockQueryCancel[*block] = append(
```

Comment:
I think this assumes the first `GetBlock` request always gets processed first, but I don't think it's true. Take @Roasbeef 's example further,
- goroutine `N_0` gets there first, followed by `N_1` closely
- `workManager` finishes querying for `N_1` first, returning an error
- `N_0` will now hang there because it was never registered a `cancelChan`, as it was the first seen request by the dispatcher.

I think we need to have `blockQueryCancel` for every block, either new or not.

Reply:
Down to the rabbit hole - does one failed `GetBlock` request means another one would also fail for the same block? I guess it depends on the implementation details of `workManager.Query`, as we may choose to send this block inv msg to different peers. I think a better way to handle this is, cache a previous success `GetBlock` result, similar to the `blockcache` in `lnd`, future PRs tho.

Reply:
I think race condition you are describing will not happen because the [newRequest](https://github.com/btcsuite/btcwallet/blob/master/chain/pruned_block_dispatcher.go#L524) method guards with the lock against successive goroutines being registered. The first block request which aquires the lock will send out the request to the internal working manager so there will be no race condition.

> I think we need to have blockQueryCancel for every block, either new or not.

We could add it but its not necessary because the initial (first) block request will use the error channel if the request fails.

Reply:
maybe I am not seeing something tho ?

Reply:
> The first block request which aquires the lock will send out the request to the internal working manager so there will be no race condition

Yeah but the lock only guarantees the registration of the block in the map. Once `d.newRequest` finishes, there's no guarantee. There's also no guarantee that ` d.workManager.Query(reqs, opts...)` will return in order, since `workManager.Query` would just put the job in a batch in an async manner.

> 

> And because we never request more than one block in what request we don't have to care about the case where we have mutliple blocks in the INV msg.

I mean even for the querying the same block multiple times we should treat each attempt differently. The mental model is different here - we think it's the same request if the block hash is the same, while in reality it's a different request, and that needs to be reflected in the design.

Reply:
ok nvm didn't realize the `reqs` are only created for the first block request, so this won't be an issue.